### PR TITLE
concurrency without model cloning

### PR DESCRIPTION
**Pull request by dtrawins**
# Support for multi threading in execution?

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by dtrawins** [comment]
@slyalin @eaidova  can you add your comments?

File: optimum/intel/openvino/modeling_decoder.py [hunk 1]
```diff
@@ -354,6 +356,18 @@ class OVModelForCausalLM(OVBaseDecoderModel, GenerationMixin):
     export_feature = "text-generation"
     auto_model_class = AutoModelForCausalLM
 
+    def generate(self, *args, **kwargs):
+        self.compile()
+        infer_context = [self.compiled_model.create_infer_request()]
+        kwargs["infer_context"] = infer_context
+        return super().generate(*args, **kwargs)
+
+    def __call__(self, *args, **kwargs):
```
Comment (slyalin) [inline optimum/intel/openvino/modeling_decoder.py:365]: @dtrawins, please explain why `__call__` should be different from `forward` behavior. `__call__` will eventually call `forward` without any added semantics on top of that. So why cannot we move this code to forward?

Comment (dtrawins) [reply optimum/intel/openvino/modeling_decoder.py]: @slyalin that is indeed not required. potentially it could be used to pass the infer_request context. Assuming we create new request in the forward method if generate method didn't pass the context, that is not required.

File: optimum/intel/openvino/modeling_decoder.py [hunk 2]
```diff
@@ -463,32 +461,41 @@ def forward(
...
+                infer_request = self.compiled_model.create_infer_request()
+        infer_request.start_async(inputs, share_inputs=True)
+        infer_request.wait()
+        logits = torch.from_numpy(infer_request.get_tensor("logits").data).to(self.device)
         if self.stateful:
             # Need a marker to differentiate the first generate iteration from the others in
             # the first condition at the function beginning above.
             # It should be something that is not None and it should be True when converted to Boolean.
-            past_key_values = ((),)
+            past_key_values = ((inputs["beam_idx"]), infer_request)
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_decoder.py:493]: not related to `past_key_values` so I don't think we should update `past_key_values` here, the resulting output will not be what it's expected for example :
```
output = model(**tokens)
pkv = output.past_key_values
```

**Review by echarlaix — COMMENTED** [review]
THanks a lot for your work @dtrawins

File: optimum/intel/openvino/modeling_base.py [hunk 3]
```diff
@@ -86,6 +86,7 @@ def __init__(
 
         self.model = model
         self.request = None
+        self.compiled_model = None
```
[the remark this answers is not in the document]
Comment (dtrawins) [reply optimum/intel/openvino/modeling_base.py:90]: It is needed to create new infer_request in the context of generate method for each concurrent thread. So far we had in the model class request attribute which was pointing to a static infer_request and can not be used to allocate new request. Generally there is a bit confusing setup when the request attribute is set to the compiled_model object in the based class but latest it is overwritten to become the infer_request. Eventually the recommendation would be to switch to using compiled_model attribute instead and create infer_requests dynamically. It was proposed to make this switch in a separate PR.

File: optimum/intel/openvino/modeling_decoder.py (hunk 2 above)
Comment (echarlaix) [reply optimum/intel/openvino/modeling_decoder.py:501]: My point is that it's not related to `past_key_values` so we shouldn't update it with beam_idx / inference execution context

Comment (dtrawins) [reply optimum/intel/openvino/modeling_decoder.py:501]: @slyalin can you add your comments here? The idea was to reused this variable for stateful models because they don't use it at all. That was the only method we found that could be used to pass the beam_idx and execution context (which includes the state data) without changing the model API. The other alternative was with using model.clone() method for each thread which would also using a separate execution context without duplicating memory consumption https://github.com/huggingface/optimum-intel/pull/564. Would cloning be better method to support concurrency in the execution? Is there some other option we are not aware of? I guess it is a bit unique situation with the stateful models in openvino so probably it is not handled in transformers lib.

Comment (slyalin) [reply optimum/intel/openvino/modeling_decoder.py:501]: > My point is that it's not related to past_key_values

Definitely it _is_ related to `past_key_values` even more than the old `((),)` value. `beam_idx` together with `infer_request` are used to track `past_key_values` for a particular sequence. Literally, `infer_request` has a model state that consists of `past_key_values` tensors, and `beam_idx` allows indirect rows reordering in that state in case of beam search. This PR just makes it more explicit than it was before and moving this attributes from the model class instance to each sequence, which allows having multiple sequences for a single model class instance.

@echarlaix, do you have a better alternative to pass these values?

Comment (slyalin) [reply optimum/intel/openvino/modeling_decoder.py:501]: > we shouldn't update this variable with beam_idx / inference execution context

I saw another example where `beam_idx` is stored directly in `past_key_values` here: https://github.com/huggingface/optimum-intel/pull/567/files#diff-96778abd26ec589f358d62af391a34fdda4f7e4734cca48b421c567f5f3042e6R591-R600. So indeed it was treated as a detail of implementation that depends on backend used for inference.

Comment (echarlaix) [reply optimum/intel/openvino/modeling_decoder.py:501]: > @echarlaix, do you have a better alternative to pass these values?

why not introducing a class inheriting from `ModelOutput` like `CausalLMOutputWithPast` https://github.com/huggingface/transformers/blob/536ea2aca234fb48c5c69769431d643b0d93b233/src/transformers/modeling_outputs.py#L678 with dedicated `beam_idx` / `inference_request` arguments ?

Comment (slyalin) [reply optimum/intel/openvino/modeling_decoder.py:501]: @echarlaix, thanks for the suggestion. But not sure that I understand the flow in this case. Do you have examples where such an inheritance is used for similar purposes?

Comment (echarlaix) [reply optimum/intel/openvino/modeling_decoder.py:501]: here for example : [BertForPreTrainingOutput](https://github.com/huggingface/transformers/blob/536ea2aca234fb48c5c69769431d643b0d93b233/src/transformers/models/bert/modeling_bert.py#L738)

Comment (dtrawins) [reply optimum/intel/openvino/modeling_decoder.py:501]: @echarlaix if we create new modelOutput data class and it is returned by the Forward method, how it could be passed back to the Forward method in the next cycle?

Comment (echarlaix) [reply optimum/intel/openvino/modeling_decoder.py:501]: can you try something like :
```python
from dataclasses import dataclass
from transformers.modeling_outputs import ModelOutput

@dataclass
class CausalLMOutputWithPast(ModelOutput):
    logits: torch.FloatTensor = None
    past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None
    beam_idx: Optional[int] = None
    inference_session = None
```



and then overwritte `_update_model_kwargs_for_generation` https://github.com/huggingface/transformers/blob/45c065109074d60c587d3e562f16531d02a422f6/src/transformers/generation/utils.py#L630 by adding somethign like :   

```python
    def _update_model_kwargs_for_generation(
        self,
        outputs: ModelOutput,
        model_kwargs: dict[str],
        is_encoder_decoder: bool = False,
        standardize_cache_format: bool = False,
    ) -> dict[str]:
        model_kwargs = super()._update_model_kwargs_for_generation(
            outputs=outputs,
            model_kwargs=model_kwargs,
            is_encoder_decoder=is_encoder_decoder,
            standardize_cache_format=standardize_cache_format,
        )

        if "beam_idx" in outputs:
            model_kwargs["beam_idx"] = outputs["beam_idx"]

        return model_kwargs
```

Comment (echarlaix) [reply optimum/intel/openvino/modeling_decoder.py:501]: Let me know if you need help on this @dtrawins

Comment (dtrawins) [reply optimum/intel/openvino/modeling_decoder.py]: @echarlaix @eaidova @slyalin Could you have a look if the latest version is passing the context fine now?
I'm not reusing past_key_values for stateful models with th generation context. There are additional fields in the `forward` output beam_idx and infer_request. Now only 9 tests is left to fix but seams unrelated to concurrency. Probably rebase from main is needed.
Anyway can one comment if beam_idx would be populated correctly. It is not defined now in reorder_caches for stateful models.

File: optimum/intel/openvino/modeling_decoder.py [hunk 4]
```diff
@@ -322,6 +340,7 @@ def normalized_config(self):
     def compile(self):
         if self.request is None:
             super().compile()
+            self.compiled_model = self.request
             self.request = self.request.create_infer_request()
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_decoder.py:344]: why not remove this
```suggestion
```
and use [self.request](https://github.com/huggingface/optimum-intel/blob/ff1d94b5522a2bdda042f02564b76686bf3add6c/optimum/intel/openvino/modeling_base.py#L408) instead of `self.compiled_model` ? (`self.request` doesn't seem to be used anywhere)

File: tests/openvino/test_modeling.py [hunk 5]
```diff
@@ -578,6 +578,72 @@ def test_compare_to_transformers(self, model_arch):
...
+                transformers_outputs = transformers_model(**tokens)
+            # Compare tensor outputs
+            self.assertTrue(torch.allclose(ov_outputs.logits, transformers_outputs.logits, atol=1e-4))
+            # self.assertTrue(False)
+
+        run_on_multiple_threads(run_ov_model, tokens_list, (transformers_model, ov_model))
+
+        del transformers_model
+        del ov_model
+        gc.collect()
```
Comment (echarlaix) [inline tests/openvino/test_modeling.py:645]: The time taken to run all tests is already non negligible so I think we should merge it with `test_compare_to_transformers` (to not duplicate steps like export)

File: tests/openvino/test_modeling.py [hunk 6]
```diff
@@ -608,6 +674,42 @@ def test_pipeline(self, model_arch):
         del model
         gc.collect()
 
+    @parameterized.expand(SUPPORTED_ARCHITECTURES)
+    def test_pipeline_multithreading(self, model_arch):
```
Comment (echarlaix) [inline tests/openvino/test_modeling.py:678]: same comment, can be merged with `test_pipeline`

File: optimum/intel/openvino/modeling_decoder.py [hunk 7]
```diff
@@ -44,6 +45,23 @@
...
+        infer_request(`openvino.runtime.InferRequest` to be reused in the generation cycles.
+        beam_idx (`torch.Tensor` beam search algorimth context for the generation using stateful models
+    """
+
+    loss: Optional[torch.FloatTensor] = None
+    logits: torch.FloatTensor = None
+    past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None
+    hidden_states: Optional[Tuple[torch.FloatTensor, ...]] = None
+    attentions: Optional[Tuple[torch.FloatTensor, ...]] = None
+    infer_request: Optional[openvino.runtime.InferRequest] = None
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_decoder.py:63]: could we rename it to something like `request` or `inference_request` ?

**Review by echarlaix — APPROVED** [review]
Looks great, thanks for iterating on this @dtrawins !

File: optimum/intel/openvino/modeling_decoder.py (hunk 2 above)
Comment (dtrawins) [reply optimum/intel/openvino/modeling_decoder.py]: What i confirmed is beam_idx was not passed correctly. The same, initial beam_idx was circulating for the whole pipeline resulting in incorrect accuracy with beam search. Somehow it was not detected by functional tests. 
Anyway my proposal is to pass the beam_idx content from reorder_caches method inside past_key_value. I tested it gives correct results and the code is in my opinion clean. The forward method returns empty past_key_values as expected for stateful models. In case someone would like to manage the pipeline for stateless models outside of transformers using just forward method, it would be still possible - beam_idx should be passed inside past_key_value and inference_request context via model_args. Anyway that is probably unlikely use case scenario. Would it be acceptable?

File: optimum/intel/openvino/modeling_decoder.py (hunk 7 above)
Comment (eaidova) [reply optimum/intel/openvino/modeling_decoder.py:63]: current name is aligned with openvino api name, so for me infer_request sounds better

Comment (echarlaix) [reply optimum/intel/openvino/modeling_decoder.py:63]: I think that'd be clearer for users who are not familiar with the openvino ecosystem, also we don't use `infer_request` anywhere in optimum-intel so was thinking about something a bit more explicit

**Comment by echarlaix** [comment]
I think we're close to merge, just waiting for couple of points above to be addressed, let me know if you need any help from my side @dtrawins (fixing conflicts / applying suggested changes)
